src/atc_tools/helpers/ExtractEventhubBody.py

```diff
@@ -0,0 +1,62 @@
+import json
+import warnings
+
+from atc.spark import Spark
+from pyspark.sql.dataframe import DataFrame
+import pyspark.sql.types as t
+import pyspark.sql.functions as f
+
+
+class ExtractEventhubBody:
+    def __init__(
+        self, json_field: str = "Body", new_column_w_extracted_body: str = None
+    ):
+        self.json_field = json_field
+        self.new_field_w_extracted_body = new_column_w_extracted_body or json_field
+
+        self._fields_same_name = self.json_field == self.new_field_w_extracted_body
+
+    def extract_json_schema(self, df: DataFrame, pretty_json: bool = False) -> str:
+        assert self.json_field in df.columns, "The body column is not in the dataframe"
+
+        # extract json data rows using name of json_field
+        json_data = df.withColumn(
+            self.json_field, f.col(self.json_field).cast(t.StringType())
+        ).rdd.map(lambda row: row.asDict()[self.json_field])
+
+        # load the json as a data frame. Schema is inferred from dara
+        json_schema_raw = str(Spark.get().read.json(json_data).schema.jsonValue())
+
+        # the returned json is non-standard. Fix it.
+        json_schema = json_schema_raw.replace("'", '"').replace(": True", ": true")
+
+        return (
+            json.dumps(json.loads(json_schema), indent=4)
+            if pretty_json
+            else json_schema
+        )
+
+    def transform_df(
+        self, df: DataFrame, keep_original_body: bool = False
+    ) -> DataFrame:
+
+        if self._fields_same_name and keep_original_body:
+            raise ValueError(
+                f"The field {self.json_field} is overwritten by the extracted json. Give extracted field new name to "
+                f"keep original body. "
+            )
+
+        # Step 1: extract the json_schema
+        json_schema = self.extract_json_schema(df)
```

Comment:
So the purpose of this entire step is to repicate the behavior of spark when reading a json stream of unknown scheama, am I understanding this right? There is even a spark setting for how many rows to read before inferring the schema.
Are you sure that you have investigated enough to be sure that this is not already supported out of the box in spark?

Reply:
No. This step extracts the schema of a single binary encoded body COLUMN.  

I think that if pyspark read the file and infers the schema , it will interpret the Body as BinaryType(). 

This class help finding out what lies beneath this binary encoded column.